File: backend/worker/handlers/bgm.py
from worker.handlers.base import BaseHandler


class BgmHandler(BaseHandler):
    async def execute(self, node_config, input_paths, output_path):
        video = input_paths["video"]
        audio = input_paths["audio"]
        volume = float(node_config.get("volume", 0.3))
        original_volume = float(node_config.get("original_volume", 1.0))
        loop = node_config.get("loop", True)
        fade_in = float(node_config.get("fade_in", 0))
        fade_out = float(node_config.get("fade_out", 0))

        probe = await self.run_ffprobe(video)
        video_duration = float(probe.get("format", {}).get("duration", 0))
        video_has_audio = any(stream.get("codec_type") == "audio" for stream in probe.get("streams", []))

        bgm_filters = [f"volume={volume}"]
        if fade_in > 0:
            bgm_filters.append(f"afade=t=in:d={fade_in}")
        if fade_out > 0 and video_duration > 0:
            fade_start = max(0, video_duration - fade_out)
            bgm_filters.append(f"afade=t=out:st={fade_start}:d={fade_out}")

        bgm_filter_chain = ",".join(bgm_filters)
        sidechain_audio_format = "aformat=sample_fmts=fltp:channel_layouts=stereo"

        input_args = ["-i", video]
        if loop:
            input_args.extend(["-stream_loop", "-1", "-i", audio])
        else:
            input_args.extend(["-i", audio])

        if video_has_audio:
            filter_complex = (
                f"[0:a]aresample=48000:async=1,{sidechain_audio_format},"
                f"volume={original_volume},asplit=2[orig_mix][orig_sidechain];"
                f"[1:a]aresample=48000:async=1,{sidechain_audio_format},{bgm_filter_chain}[bgm];"
                "[bgm][orig_sidechain]sidechaincompress=threshold=0.03:ratio=8:attack=200:release=800[ducked];"
                "[orig_mix][ducked]amix=inputs=2:duration=first:normalize=0[mix];"
                "[mix]loudnorm=I=-16:LRA=11:TP=-1.5[a]"
            )
        else:
            filter_complex = f"[1:a]aresample=48000:async=1,{bgm_filter_chain},loudnorm=I=-16:LRA=11:TP=-1.5[a]"

        args = input_args + [
            "-filter_complex", filter_complex,
            "-map", "0:v", "-map", "[a]",
            "-c:v", "copy",
            "-c:a", "aac",
            "-ar", "48000",
            "-ac", "2",
            "-shortest",
            output_path,
        ]
        await self.run_ffmpeg(args)

File: internal/worker/handlers/bgm.go
package handlers

import (
	"context"
	"errors"
	"strings"

	vpffmpeg "github.com/Ctwqk/videoprocess/internal/worker/ffmpeg"
)

type BgmHandler struct {
	Runner vpffmpeg.Runner
}

func (h BgmHandler) NodeType() string {
	return "bgm"
}

func BgmArgs(videoPath string, audioPath string, outputPath string, config map[string]any, videoProbe probeSummary) []string {
	volume := floatValue(config["volume"], 0.3)
	originalVolume := floatValue(config["original_volume"], 1.0)
	loop := truthyValue(config["loop"], true)
	fadeIn := floatValue(config["fade_in"], 0)
	fadeOut := floatValue(config["fade_out"], 0)

	bgmFilters := []string{"volume=" + formatFloat(volume)}
	if fadeIn > 0 {
		bgmFilters = append(bgmFilters, "afade=t=in:d="+formatFloat(fadeIn))
	}
	if fadeOut > 0 && videoProbe.Duration > 0 {
		fadeStart := videoProbe.Duration - fadeOut
		if fadeStart < 0 {
			fadeStart = 0
		}
		bgmFilters = append(bgmFilters, "afade=t=out:st="+formatFloat(fadeStart)+":d="+formatFloat(fadeOut))
	}

	bgmFilterChain := strings.Join(bgmFilters, ",")
	sidechainAudioFormat := "aformat=sample_fmts=fltp:channel_layouts=stereo"

	inputArgs := []string{"-i", videoPath}
	if loop {
		inputArgs = append(inputArgs, "-stream_loop", "-1", "-i", audioPath)
	} else {
		inputArgs = append(inputArgs, "-i", audioPath)
	}

	var filterComplex string
	if videoProbe.HasAudio {
		filterComplex = "[0:a]aresample=48000:async=1," + sidechainAudioFormat + "," +
			"volume=" + formatFloat(originalVolume) + ",asplit=2[orig_mix][orig_sidechain];" +
			"[1:a]aresample=48000:async=1," + sidechainAudioFormat + "," + bgmFilterChain + "[bgm];" +
			"[bgm][orig_sidechain]sidechaincompress=threshold=0.03:ratio=8:attack=200:release=800[ducked];" +
			"[orig_mix][ducked]amix=inputs=2:duration=first:normalize=0[mix];" +
			"[mix]loudnorm=I=-16:LRA=11:TP=-1.5[a]"
	} else {
		filterComplex = "[1:a]aresample=48000:async=1," + bgmFilterChain + ",loudnorm=I=-16:LRA=11:TP=-1.5[a]"
	}

	args := append([]string{}, inputArgs...)
	args = append(args,
		"-filter_complex", filterComplex,
		"-map", "0:v", "-map", "[a]",
		"-c:v", "copy",
		"-c:a", "aac",
		"-ar", "48000",
		"-ac", "2",
		"-shortest",
		outputPath,
	)
	return args
}

func (h BgmHandler) Execute(ctx context.Context, inputPaths map[string]string, outputPath string, config map[string]any) (map[string]any, error) {
	videoPath := inputPaths["video"]
	if videoPath == "" {
		return nil, errors.New("missing input path on video port")
	}
	audioPath := inputPaths["audio"]
	if audioPath == "" {
		return nil, errors.New("missing input path on audio port")
	}
	videoProbe, err := probePath(ctx, h.Runner, videoPath)
	if err != nil {
		return nil, err
	}
	if err := runFFmpeg(ctx, h.Runner, BgmArgs(videoPath, audioPath, outputPath, config, videoProbe)); err != nil {
		return nil, err
	}
	return map[string]any{}, nil
}
